### PR TITLE
feat: add segment tracking

### PR DESCRIPTION
**Pull request by mikemurray**
Issue: #58 
Impact: **minor**
Type: **feat**

## Issue

Add an event tracking API to the app and have that data be sent to Segment.

## Solution

- Add segment scripts using th next js example https://github.com/zeit/next.js/tree/canary/examples/with-segment-analytics
- Use the NYT react tracking lib https://github.com/NYTimes/react-tracking
- Create a custom dispatch and HOC decorators under `lib/tracking`
- Add a `config/analytics` directory with the segment tracking implementation as well as an example for creating more
- Add an ENV var `SEGMENT_ANALYTICS_WRITE_KEY` to `.env.example` for easy configuration
- Create sample tracking HOC's for tracking a product and product list, to be fully implemented in the future by #56 and #57 
- Added docs to outlining how to set up. use, and customize tracking.
- Add the `_app.js` override component with the root level tracking

## Breaking

- `this.props.url` is now `this.props.router`

## Testing

1. Create a segment account and get the API key
2. Start the app and navigate around
3. In the segment dashboard under `sources > javascript > debugger` see events being tracked

### CODE REVIEW

File: src/pages/_document.js [hunk 1]
```diff
@@ -50,6 +51,19 @@ class HTMLDocument extends Document {
     };
   };
 
+  renderAnalyticsSnippet() {
+    const opts = {
+      apiKey: rootMobxStores.uiStore.appConfig.publicRuntimeConfig.segmentAnalyticsWriteKey,
+      page: true // Set this to `false` if you want to manually fire `analytics.page()` from within your pages.
+    };
+
+    if (process.env.NODE_ENV === "development") {
```
[the remark this answers is not in the document]
Comment (mikemurray) [reply src/pages/_document.js]: @ticean You're purposing having a`SEGMENT_SNIPPET_ENV=max` or something to that extent just so it's not tied to the single node NODE_ENV?

In this case of this conditional, it's switching between the minified and non-minified versions of the segment snippet. Would control over that be helpful, or add more configuration options than a user might not be aware of the ramifications of?

File: src/lib/tracking/withTracking.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+export default (Component) => (
+  @track({}, { dispatch })
+  class withTracking extends React.Component {
+    render() {
+      return (
+        <Component {...this.props} />
+      );
+    }
+  }
+);
```
Comment (aldeed) [inline src/lib/tracking/withTracking.js]: To have fewer unnecessary layers, I'd prefer to use the NYT recommended approach of passing dispatch just once, in the top layer of the app: https://github.com/NYTimes/react-tracking#custom-optionsdispatch-for-tracking-data

Then `withTracking` should be unnecessary.

Was there a reason this approach doesn't work?

File: docs/tracking-events.md [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+  return "STRING_SCRIPT";
+}
+```
+
+### Step 4: Testing
+
+You should now be able to send tracking data to the provider of your choice.
+
+
+## Add tracking to a page
```
Comment (machikoyasuda) [inline docs/tracking-events.md]: Add a page view event

File: docs/tracking-events.md [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+})
+class Page extends Component {
+  render() {
+    return <div>{"Page"}</div>;
+  }
+}
+
+```
+
+## Tracking events inside a component
```
Comment (machikoyasuda) [inline docs/tracking-events.md]: Add events inside a component

File: docs/tracking-events.md [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+        <button onClick={this.handleClick}>{"Track Event 1"}</button>
+        <button onClick={this.handleOtherClick}>{"Track Event 2"}</button>
+      </div>
+    );
+  }
+}
+
+```
+
+## Track product list viewed event
```
Comment (machikoyasuda) [inline docs/tracking-events.md]: Add a Product List Viewed event

File: docs/tracking-events.md [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+          <div>{catalogItem.node.title}</div>
+        })}
+      </div>
+    );
+  }
+}
+
+```
+
+## Track product viewed event
```
Comment (machikoyasuda) [inline docs/tracking-events.md]: Add a Product Viewed event

File: docs/tracking-events.md [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    return (
+      <button onClick={this.handleClick}>{"Track Event 1"}</button>
+      <button onClick={this.handleOtherClick}>{"Track Event 2"}</button>
+    );
+  }
+}
+
+```
+
+## Tracking events inside a component and on page load
```
Comment (machikoyasuda) [inline docs/tracking-events.md]: Add events on page load and inside a component

File: .env.example [hunk 8]
```diff
@@ -4,4 +4,5 @@ EXTERNAL_GRAPHQL_URL=http://localhost:3030/graphql-alpha
 EXTERNAL_ASSETS_URL=http://localhost:3000
 PLACEHOLDER_IMAGE_URL_GALLERY=/resources/placeholder.gif
 PLACEHOLDER_IMAGE_URL_GRID=/resources/placeholder.gif
+SEGMENT_ANALYTICS_ENV=development
```
[the remark this answers is not in the document]
Comment (mikemurray) [reply .env.example]: Gotcha!

File: config/analytics/segment.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+ * @name dispatch
+ * @ignore
+ * @param {Object} data Arguments supplied by tracking library
+ * @param {String} data.action String value used as the first param to analytics.track()
+ * @returns {undefined} No Return
+ */
+export function dispatch(data) {
+  // Workaround for not being able to use object rest spread
+  const { action } = data;
+  delete data.action;
```
[the remark this answers is not in the document]
Comment (mikemurray) [reply config/analytics/segment.js]: I can use `Object.assign({}, data)` to avoid mutating the param.

**Comment by mikemurray** [comment]
@aldeed Ready for a re-review

**Pull request merged**